google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,392 @@
+                    resource_id=rid,
+                    resource_type=resource_type)
+
+                self.resource_rules_map[apply_to][gcp_resource].add(
+                    rule)
+
+    def get_resource_rules(self, applies_to):
+        """Get all the rules for the resource "applies_to".
+        Args:
+            applies_to (str): The name of the resource
+
+        Returns:
+           defaultdict:  A list of ResourceRules.
+        """
+        return self.resource_rules_map[applies_to]
+
+
+class Rule(object):
+    """Rule properties from the rule definition file.
+    Also finds violations.
+    """
+
+    def __init__(self, rule_name, rule_index, min_retention, max_retention):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule
+            rule_index (int): The index of the rule
+            min_retention(int): minimum value of the age in lifecycle
+            max_retention(int): maximum value of the age in lifecycle
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.min_retention = min_retention
+        self.max_retention = max_retention
+
+    def generate_bucket_violation(self, bucket):
+        """generate a violation.
+
+        Args:
+            bucket (Bucket): The bucket that triggers the violation
+        Returns:
+            RuleViolation: The violation
+        """
+        data_lifecycle = bucket.get_lifecycle_rule()
+        data_lifecycle_str = json.dumps(data_lifecycle)
+
+        return RuleViolation(
+            resource_name=bucket.name,
+            resource_id=bucket.id,
+            resource_type=bucket.type,
+            full_name=bucket.full_name,
+            rule_name=self.rule_name,
+            rule_index=self.rule_index,
+            violation_type=VIOLATION_TYPE,
+            violation_data=data_lifecycle_str,
+            resource_data=bucket.data,
+        )
+
+    def find_violations(self, res):
+        """Get a generator for violations
+
+        Args:
+            res (Resource): A class derived from Resource
+        Returns:
+            Generator: All violations of the resource breaking the rule.
+        """
+
+        if res.type == 'bucket':
+            return self.find_violations_in_bucket(res)
+        return iter(())
+
+    def bucket_max_retention_violation(self, bucket):
+        """Get a generator for violations especially for maximum retention
+           It only supports bucket for now, and will work on generalizing
+           in future PRs.
+
+        Args:
+            bucket (bucket): Find violation from the bucket
+        Yields:
+            RuleViolation: All max violations of the bucket breaking the rule.
+        """
+        if self.max_retention is None:
+            return
+
+        # There should be a condition which guarantees to delete data
+        bucket_lifecycle = bucket.get_lifecycle_rule()
+        if not bucket_lifecycle:
```

Comment:
If you invert this condition to just `if bucket_lifecycle` with the `for` loop inside, then you can have the yield of violation only place, at the end of the function